### PR TITLE
Deprecate add_to_queue and get_from_queue from Lightning Module

### LINKED ISSUE #8940: Deprecate add_to_queue / get_from_queue
## 🚀 Feature

Deprecate these two functions from the lightning interface:
https://github.com/PyTorchLightning/pytorch-lightning/blob/4b6aaeeae3c53c45cf31a7fc2cf90133acfe5d12/pytorch_lightning/core/lightning.py#L1989-L2014

### Motivation

Both of these hooks are hyper-specific to transferring data in a spawned manner. They are not related to ML code at all and should not be on the lightning module interface

more discussion here: https://docs.google.com/document/d/1xHU7-iQSpp9KJTjI3As2EM0mfNHHr37WZYpDpwLkivA/edit?disco=AAAANaZ8IYw

This is part of a larger initiative to separate Trainer and lightning module #7315, as part of the Lightning Architecture Review #7740

### Pitch

* Deprecate this method in v1.5
* Remove this method in v1.7

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by daniellepintz**
## What does this PR do?

Fixes #8940

Deprecates `add_to_queue`/`get_from_queue` from Lightning Module. To address #7671 in v1.7 we will put `trainer.callback_metrics` directly on the queue in `__transfer_distrib_spawn_state_on_fit_end` and get it from the queue in `post_dispatch`.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/configuration_validator.py [hunk 1]
```diff
@@ -153,3 +154,20 @@ def __check_training_step_requires_dataloader_iter(self, model: "pl.LightningMod
...
                     "is incompatible with `truncated_bptt_steps > 0`."
                 )
+
+    def _check_add_get_queue(self, model: "pl.LightningModule"):
+        r"""
+        Checks if add_to_queue or get_from_queue is overriden and sends a deprecation warning.
+
+        Args:
+            model: The lightning module
+
```
Comment (awaelchli) [inline pytorch_lightning/trainer/configuration_validator.py:164]:
```suggestion
```

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 2]
```diff
@@ -213,13 +213,18 @@ def new_process(self, process_idx: int, trainer: "pl.Trainer", mp_queue: SimpleQ
...
         best_path = self.mp_queue.get()
         last_path = self.mp_queue.get()
         self._results = self.mp_queue.get()
+        # TODO(@daniellepintz): add `trainer.callback_metrics = self.mp_queue.get()` in v1.7
+
         # get the `callback_metrics` and set it to the trainer
         # only in case the user does not override it.
+
+        # TODO(@daniellepintz): remove in v1.7
         self.lightning_module.get_from_queue(self.mp_queue)
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/ddp_spawn.py:228]: to add to @tchaton request, I believe we should let the user override these hooks on the ddpspawn plugin during this deprecation phase (when they see the warning), so they will be ready when the removal takes place in 1.7.

File: CHANGELOG.md [hunk 3]
```diff
@@ -162,6 +162,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 
 
+- Deprecated `add_to_queue`, `get_from_queue` from Lightning Module ([#9126](https://github.com/PyTorchLightning/pytorch-lightning/pull/9126))
```
Comment (awaelchli) [inline CHANGELOG.md:165]:
```suggestion
- Deprecated `add_to_queue`, `get_from_queue` from Lightning Module in favor of overriding the same in the DDPSpawnPlugins ([#9126](https://github.com/PyTorchLightning/pytorch-lightning/pull/9126))
```

File: pytorch_lightning/trainer/configuration_validator.py [hunk 4]
```diff
@@ -153,3 +154,20 @@ def __check_training_step_requires_dataloader_iter(self, model: "pl.LightningMod
...
+        """
+        if is_overridden("add_to_queue", model):
+            rank_zero_deprecation(
+                "The `LightningModule.add_to_queue` method was deprecated in v1.5 and will be removed in v1.7."
```
Comment (awaelchli) [inline pytorch_lightning/trainer/configuration_validator.py:168]: let's mention that user can override the same hook on the plugin?

File: CHANGELOG.md (hunk 3 above)
Comment (daniellepintz) [reply CHANGELOG.md:165]: actually I am proposing we deprecate these functions and dont offer an alternative, see my comment in https://github.com/PyTorchLightning/pytorch-lightning/pull/9118#issuecomment-907429486

do you think that's possible?

Comment (awaelchli) [reply CHANGELOG.md:165]: these methods were introduced in 1.4.4. If they are undocumented (it seems not mentioned in changelog) then maybe it's ok to remove them but generally speaking I think we cannot just remove API without offering an alternatie.

**Comment by daniellepintz** [comment]
Closing in favor of #9118
